File: RayTracer/Light/DirectionalLight.cs
using RayTracer.Common;
using RayTracer.Material;
using RayTracer.Shape;
using RayTracer.Tracer;
using System;
using System.Collections.Generic;
using RayTracer.BVH;

namespace RayTracer.Lighting
{
    [Serializable]
    public class DirectionalLight : Light
    {


        public DirectionalLight(Vec3 direction, MyColor color)
        {
            Direction = direction.Normalize();
            Color = color;

        }
        public DirectionalLight(float[] param)
            : this(new Vec3(param[0], param[1], param[2]), new MyColor(param[3], param[4], param[5]))
        {
        }

        public Vec3 Direction
        {
            get;
            set;
        }

        public override Vec3 GetPointToLight(Point3 point)
        {
            return new Vec3(Direction * -1);
        }

        public override bool IsEffective(Point3 point, Container bvh)
        {
            Ray shadowRay = new Ray(point, (Direction*-1));
            if (bvh.IsIntersecting(shadowRay))
            {
                if (bvh.Geo != null)
                {
                    shadowRay.TransformInv(bvh.Geo.Trans);
                    if (bvh.Geo.IsIntersecting(shadowRay))
                            return false;
                }
                else
                {
                    foreach (Container bin in bvh.Childs)
                        if (!IsEffective(point, bin)) return false;
                }
            }

            return true;
        }
        

        public override float GetAttValue(Point3 point, Attenuation attenuation)
        {
            return 1;
        }


    }
}

File: RayTracer - BVH - CUDA/RayTracerCUDA/DirectionalLight.cpp
#include "DirectionalLight.h"



DirectionalLight::DirectionalLight()
{
}


DirectionalLight::~DirectionalLight()
{
}

bool DirectionalLight::isEffective(Vec3 & point, std::shared_ptr< Container> bvh)
{
	Ray shadowRay = Ray(point, (*dir * -1.f));

	std::vector<std::shared_ptr< Container>> bins;

	bins.push_back(bvh);
	while (bins.size() > 0)
	{
		std::shared_ptr< Container> currBin = bins.back();
		bins.pop_back();
		if (currBin->isIntersecting(shadowRay))
		{
			if (currBin->geo != nullptr)
			{
				
				if (currBin->geo->isIntersecting(shadowRay))
					return false;
				
			}
			else
			{
				bins.push_back(currBin->rChild);
				bins.push_back(currBin->lChild);
			}
		}
	}

	return true;
}

Vec3 DirectionalLight::getPointToLight(const Vec3 & point)
{
	return Vec3(*dir * -1);
}
